src/core/models.py

```diff
@@ -20,34 +21,64 @@ class StateBase:
     """Base state object."""
 
     def __init__(
-        self, relation: Relation | None, component: Unit | Application, substrate: SUBSTRATES
+        self,
+        relation: Relation | None,
+        data_interface: Data,
+        component: Unit | Application,
+        substrate: SUBSTRATES,
     ):
         self.relation = relation
+        self.data_interface = data_interface
         self.component = component
         self.substrate = substrate
 
-    @property
-    def relation_data(self) -> MutableMapping[str, str]:
-        """The raw relation data."""
-        if not self.relation:
-            return {}
+    def update(self, items: dict[str, str]) -> None:
+        """Changes the state."""
+        raise NotImplementedError
+
+    def data(self) -> MutableMapping:
+        """Data representing the state."""
+        raise NotImplementedError
+
+
+class RelationState(StateBase):
+    """Base state object."""
 
-        return self.relation.data[self.component]
+    def __init__(
+        self,
+        relation: Relation,
+        data_interface: Data,
+        component: Unit | Application,
+        substrate: SUBSTRATES,
+    ):
+        super().__init__(relation, data_interface, component, substrate)
+        # Redundant definition as lint can't resolve that super's relation may be None
+        self.relation = relation
+        self.relation_data = cast(dict[str, str], self.data_interface.as_dict(self.relation.id))
```

Comment:
Address https://github.com/canonical/zookeeper-operator/pull/129#discussion_r1555887269

Reply:
Removed for consistency